### PR TITLE
Reapply `torch.compile` in Fabric.setup()

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Alternative implementation to #19192

---

Part of #17250

When you do

```py
model = torch.compile(orig_model)
model = fabric.setup(model)
```

today, then you end up with
```py
FabricModule(StrategyWrapper(OptimizedModule(orig_model)))
```
where StrategyWrapper represents the wrapper applied by the strategy, e.g., FSDP, DDP, DeepSpeedEngine, etc.

With the new feature enabled in this PR, you will now get:

```py
FabricModule(OptimizedModule(StrategyWrapper(orig_model)))
```
Note how the optimized module gets applied outside the strategy wrapper, but inside the fabric module router.
It is important to have the fabric module outside the optimized module, because it contains only convenience routing and precision management that does not need to be compiled (nor can it be without graph breaks atm). At the same time, the FSDP/DDP specific operations can still be captured properly by torch compile, and how it is intended to be wrapped in PyTorch.

This feature is marked as **experimental** because we first need to test its application in the wild.

Follow-up work:
- Verify it in the wild
- Decide on the default
- Document it in the compile-guide
- Do the same in Lightning Trainer

----
📚 Documentation preview 📚: https://pytorch-lightning--19280.org.readthedocs.build/en/19280/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/fabric/wrappers.py [hunk 1]
```diff
@@ -322,5 +349,28 @@ def is_wrapped(obj: object) -> bool:
...
+        compiled_model = compile_fn(model, **kwargs)
+        if isinstance(model, nn.Module):
+            compiled_model._compile_kwargs = deepcopy(kwargs)
+        return compiled_model
+
+    return _capture
+
+
+if _TORCH_GREATER_EQUAL_2_0:
+    torch.compile = _capture_compile_kwargs(torch.compile)
```
Comment (awaelchli) [inline src/lightning/fabric/wrappers.py:376]: Maybe this should go at the top of `lightning/fabric/fabric.py`?

File: src/lightning/fabric/wrappers.py [hunk 2]
```diff
@@ -322,5 +349,28 @@ def is_wrapped(obj: object) -> bool:
         obj: The object to test.
 
     """
-    obj = _unwrap_compiled(obj)
+    obj, _ = _unwrap_compiled(obj)
     return isinstance(obj, (_FabricModule, _FabricOptimizer, _FabricDataLoader))
+
+
+def _capture_compile_kwargs(compile_fn: Callable) -> Callable:
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:356]: To be exhaustive, you would also need to capture the global dynamo/inductor state at this point in time and restore it when compile is actually applied:

https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/config.py
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/config.py

File: src/lightning/fabric/wrappers.py [hunk 3]
```diff
@@ -296,19 +314,28 @@ def _unwrap(
...
-        return obj
+        # obj can't be an `OptimizedModule` anyway
+        return obj, None
+
     from torch._dynamo import OptimizedModule
 
     if isinstance(obj, OptimizedModule):
-        return obj._orig_mod
-    return obj
+        return obj._orig_mod, getattr(obj, "_compile_kwargs", None)
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:330]: If `_compile_kwargs` doesn't exist, this should raise an error asking to import Fabric before torch.compile is used

File: src/lightning/fabric/wrappers.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/fabric/wrappers.py:361]: I would give a) a week and if nothing happens, go with c) and add comments about this limitation/unknown

Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:367]: I added comments about the limitation in the function that extracts the arguments.

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**